cypress-io/cypress
# fix: iterative snapshot generation with a new intermediate deferred or norewrite dependency

- Closes #22989

### User facing changelog


n/a

### Additional details


Currently, iterative snapshot generation fails under the following scenario:

1. Generate a snapshot
2. Add a require for a new deferred or norewrite file that is not a leaf.
3. Regenerate the snapshot using the snapshot metadata from step 1 as a starting point

The problem is in the logic that determines the next phase of snapshot doctoring. If all leaves have been accounted for but there is an unaccounted for intermediate dependency a next doctoring phase will not be found. I am refactoring the code to ensure that this edge case also accounts for finding the appropriate next phase of doctoring.

Note: There is still a scenario that will fail snapshot generation. In this case snapshot generation will fail if a healthy dependency changes to a deferred or norewrite dependency. I am leaving a skipped test in the PR for that scenario, but I did some analysis and there were no instances of this happening since we released 10.0 so it likely will not occur that frequently. I logged https://github.com/cypress-io/cypress/issues/23690 to address this later. Until that is addressed, snapshots will have to be generated from scratch if that happens. I will add instructions for how to do that on my subsequent PR that incorporates snapshot generation into the monorepo.

### Steps to test


I added a test case as part of this PR, you can verify this works by seeing the test case fail without my changes and then ensure that it succeeds with my changes.

### How has the user experience changed?


n/a

### PR Tasks


- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
